src/OrchardCore.Modules/OrchardCore.Workflows/WorkflowPruning/Services/WorkflowPruningManager.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using OrchardCore.Modules;
+using OrchardCore.Settings;
+using OrchardCore.Workflows.Indexes;
+using OrchardCore.Workflows.Models;
+using OrchardCore.Workflows.WorkflowPruning.Models;
+using YesSql;
+using YesSql.Services;
+
+namespace OrchardCore.Workflows.WorkflowPruning.Services;
+
+public class WorkflowPruningManager : IWorkflowPruningManager
+{
+    private readonly ISiteService _siteService;
+    private readonly ISession _session;
+    private readonly IClock _clock;
+
+    public WorkflowPruningManager(ISiteService siteService, ISession session, IClock clock)
+    {
+        _siteService = siteService;
+        _session = session;
+        _clock = clock;
+    }
+
+    public async Task<int> PruneWorkflowInstancesAsync(TimeSpan retentionPeriod)
+    {
+        var dateThreshold = _clock.UtcNow.AddDays(1) - retentionPeriod;
+        var settings = await _siteService.GetSettingsAsync<WorkflowPruningSettings>();
+
+        settings.Statuses ??=
+        [
+            WorkflowStatus.Idle,
+            WorkflowStatus.Starting,
+            WorkflowStatus.Resuming,
+            WorkflowStatus.Executing,
+            WorkflowStatus.Halted,
+            WorkflowStatus.Finished,
+            WorkflowStatus.Faulted,
+            WorkflowStatus.Aborted
+        ];
+
+        if (settings.Statuses.Length <= 0)
+        {
+            return 0;
+        }
+
+        var statuses = settings.Statuses.Select(x => (int)x).ToArray();
+        var workflowInstances = await _session
+            .Query<Workflow, WorkflowIndex>(x => x.WorkflowStatus.IsIn(statuses) && x.CreatedUtc <= dateThreshold)
+            .ListAsync();
+
+        var total = 0;
+        foreach (var item in workflowInstances)
+        {
+            _session.Delete(item);
+            total++;
+        }
```

Comment:
hmm, yeah we did batching on yessql some time ago, which kinda solved that.

It didn't cause us a problem when we released this code 6 months ago, I would hesitate a guess that we were killing of 100K+ workflows on that initial run, but we have reasonable performance on the boxes too.

May pay to limit it to say 10,000 per run, and they will clean up over a matter of days, rather than that original hit

Reply:
Ah, we actually have batching! https://github.com/sebastienros/yessql/issues/310 I totally forgot about this. The default page size for that [is 500](https://github.com/sebastienros/yessql/blob/1ff3fd3894a6b591de8c7f9ee3301eb51ccf8972/src/YesSql.Core/Configuration.cs#L22), so the 860k workflows I mentioned under https://github.com/OrchardCMS/OrchardCore/issues/16447 should be less than 2000 queries. Even if these take like 5 ms each, which they shouldn't by far, this will only take 10s, which should be safe.

I'll still do a test with the real DB once the export is ready, to see if there are any surprises with some actual data. I'll report back.

Reply:
ok, let me know what you find, and I'll get another pr going

Reply:
I now checked, and it took 90s with a local SQL Server instance to delete ~870k Workflows. Apparently, my mock workflows were too simple to provide an accurate model. And it didn't succeed even locally, despite the connection timeout being larger in the connstring (I don't know actually what happens, because there's no exception, control just never goes to the next line after [here](https://github.com/OrchardCMS/OrchardCore/blob/5c45566274b96dae775af260c3b3214d418ee2e1/src/OrchardCore.Modules/OrchardCore.Workflows/WorkflowPruning/Services/WorkflowPruningBackgroundTask.cs#L48)).

So, the DB connection will definitely time out for this in prod (where it'll take at least 10x or 100x as long due to the DB server being not local but over the network).

Adding `Take(50000)` brings down the runtime of the deletion to ~4, but it still gets stuck on `LoadSiteSettingsAsync` as mentioned above, 10k worked f). So, limiting the number here can definitely be a workaround. I'd add an `IOptions` config and use something lower like 5k (which takes <1s) as the default. Then, we can advise people to increase this if their hosting environment allows, and to temporarily increase the background task frequency when they first deploy this to an existing site to get rid of the garbage faster (running this once every minute shouldn't be an issue for any proper server, really, the problem is just the timeout of doing too much at once)

Reply:
BTW let it run with 10k to completion, it worked well.